lib/model/stream.dart

```diff
@@ -0,0 +1,230 @@
+import '../api/model/events.dart';
+import '../api/model/initial_snapshot.dart';
+import '../api/model/model.dart';
+
+/// The portion of [PerAccountStore] for streams, topics, and stuff about them.
+///
+/// This type is useful for expressing the needs of other parts of the
+/// implementation of [PerAccountStore], to avoid circularity.
+///
+/// The data structures described here are implemented at [StreamStoreImpl].
+mixin StreamStore {
+  Map<int, ZulipStream> get streams;
+  Map<String, ZulipStream> get streamsByName;
+  Map<int, Subscription> get subscriptions;
+  UserTopicVisibilityPolicy topicVisibilityPolicy(int streamId, String topic);
+
+  /// Whether this topic should appear when already focusing on its stream.
+  ///
+  /// This is determined purely by the user's visibility policy for the topic.
+  ///
+  /// This function is appropriate for muting calculations in UI contexts that
+  /// are already specific to a stream: for example the stream's unread count,
+  /// or the message list in the stream's narrow.
+  ///
+  /// For UI contexts that are not specific to a particular stream, see
+  /// [isTopicVisible].
+  bool isTopicVisibleInStream(int streamId, String topic) {
+    switch (topicVisibilityPolicy(streamId, topic)) {
+      case UserTopicVisibilityPolicy.none:
+        return true;
+      case UserTopicVisibilityPolicy.muted:
+        return false;
+      case UserTopicVisibilityPolicy.unmuted:
+      case UserTopicVisibilityPolicy.followed:
+        return true;
+      case UserTopicVisibilityPolicy.unknown:
+        assert(false);
+        return true;
+    }
+  }
+
+  /// Whether this topic should appear when not specifically focusing
+  /// on this stream.
+  ///
+  /// This takes into account the user's visibility policy for the stream
+  /// overall, as well as their policy for this topic.
+  ///
+  /// For UI contexts that are specific to a particular stream, see
+  /// [isTopicVisibleInStream].
+  bool isTopicVisible(int streamId, String topic) {
+    switch (topicVisibilityPolicy(streamId, topic)) {
+      case UserTopicVisibilityPolicy.none:
+        switch (subscriptions[streamId]?.isMuted) {
+          case false: return true;
+          case true:  return false;
+          case null:  return false; // not subscribed; treat like muted
```

Comment:
The test failures are because I need to make this line more conditional — in some contexts (like the inbox), we want to treat a not-subscribed stream the same as a muted stream, but in others (like the message list), we want to treat it the same as a subscribed and not muted stream.

Marking as draft until I fix that, but in the meantime the rest of the code is ready to review.